state/license.go

```diff
@@ -0,0 +1,167 @@
+package state
+
+import (
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/kong/deck/utils"
+)
+
+const (
+	licenseTableName = "license"
+)
+
+var licenseTableSchema = &memdb.TableSchema{
+	Name: licenseTableName,
+	Indexes: map[string]*memdb.IndexSchema{
+		"id": {
+			Name:    "id",
+			Unique:  true,
+			Indexer: &memdb.StringFieldIndex{Field: "ID"},
+		},
+		all: allIndex,
+	},
+}
+
+// LicensesCollection stores and indexes Kong Licenses.
+type LicensesCollection collection
+
+// Add adds a license to the collection.
+// license.ID should not be nil else an error is thrown.
+func (k *LicensesCollection) Add(license License) error {
+	if utils.Empty(license.ID) {
+		return errIDRequired
+	}
+	txn := k.db.Txn(true)
+	defer txn.Abort()
+
+	var searchBy []string
+	searchBy = append(searchBy, *license.ID)
+	_, err := getLicense(txn, searchBy...)
+	if err == nil {
+		return fmt.Errorf("inserting license %v: %w", license.Console(), ErrAlreadyExists)
+	} else if err != ErrNotFound {
+		return err
+	}
```

Comment:
I believe if this was consistent with what's used in `LicensesCollection.Get()` it would be a bit more readable.

```suggestion
	if err == nil {
		if err == ErrNotFound {
			return err
		}
		return fmt.Errorf("inserting license %v: %w", license.Console(), ErrAlreadyExists)
	}
```

Secondly, does that actually work? Don't we need to use `errors.Is()`?

Reply:
Also, I think the `if err == nil` check should actually be `if err != nil`.

Reply:
Reverted this after initially accepting it. Unless I'm missing something, nesting the `if err == ErrNotFound` under the `if err == nil` means we never hit the former case, correct? Was there something else intended here?

This shouldn't match `Get()`, since the good and bad conditions here are the reverse of `Get()`: not found is what we want to see, whereas a successful `getLicense()` response is a failure condition.

`Get()`, however, looks like it can be simplified, since `getLicense()` returns its own `nil` license value on errors, and we don't need an `if err == ErrNotFound { return ErrNotFound }` if we're going to return any error after anyway.

The equivalency check only needs `Is()` or `As()` if we wrap it, correct? We don't; there's only the one possible string for `ErrNotFound`.

Reply:
> The equivalency check only needs `Is()` or `As()` if we wrap it, correct? We don't; there's only the one possible string for `ErrNotFound`.

Yes, but using `Is()` or `As()` will be future proof in case that would change. 🤷